### ISSUE TITLE
Friendly format JSON Map files

### ISSUE THREAD

**Issue by griest024** [bjorn/tiled#1684]
Currently Tiled puts all the data for a layer on a single line in a json file. This makes it very inconvenient for version control as git VCs by line. Could there be an option to format the JSON is a more human and git friendly way?

**Comment by bjorn** [comment]
I don't know of any JSON library that allows this, so I think this would mean writing our own JSON writer. It would not be able to use intermediate representations based on `QVariantList` or `QJsonArray`, since such data types do not allow for including the needed formatting information.

Essentially, it would need to work much like the Lua export, which uses a custom `LuaTableWriter`. And to avoid maintaining a lot of additional code, it would be interesting to look at whether the `LuaTableWriter` could be made more generic such that it can also format JSON documents.

**Comment by griest024** [comment]
Yeah I don't think its worth writing a custom writer just for this. I find it odd that there's no C++ lib that can do it. For example, the ruby stdlib has a [pretty_generate](https://ruby-doc.org/stdlib-2.4.1/libdoc/json/rdoc/JSON.html#method-i-pretty_generate) methos and there are plenty of npm packages that can do it. Does C/C++/Qt just have poor JSON support in general?

**Comment by griest024** [comment]
Looks like [this library](https://github.com/nlohmann/json) can do it with use of the [indent parameter](https://nlohmann.github.io/json/classnlohmann_1_1basic__json_a5319dc1bb9dfe19ce7ff559aaded3422.html#a5319dc1bb9dfe19ce7ff559aaded3422)

**Comment by termhn** [comment]
As a workaround for now, it's fairly easy to open a file in a text editor and do an autoformat, and you can probably even automate it if you think that will really save you time with eslint transforms or similar.

**Comment by bjorn** [comment]
> Looks like this library can do it with use of the indent parameter.

It seems that it has the option of putting each element on its own line. But that doesn't get you very far in terms of readability in case of tile layer data, and it increases the file size quite a bit due to all the indentation.

**Comment by bjorn** [comment]
> I find it odd that there's no C++ lib that can do it. For example, the ruby stdlib has a pretty_generate method and there are plenty of npm packages that can do it. Does C++/Qt just have poor JSON support in general?

I don't think `pretty_generate` can actually do what you're asking in your initial post? If so, I'd be interested to see how. Or to see how to do it with an npm package.

If you're only asking for each tile number to be printed on its own line, then this could be done with Qt as well. I just fail to see how that's really more helpful.

**Comment by griest024** [comment]
> If you're only asking for each tile number to be printed on its own line, then this could be done with Qt as well. I just fail to see how that's really more helpful.

Yes this is all I really need. Currently, with all layer data on one line, if I make a change to a map, then commit and push that change then if my team member makes a change to that same layer, it will show up as a conflict, even if we didn't change the same tile.

If each tile number had its own line then git would only show a conflict if my team member and I actually changed the *exact same* tile.

It would certainly increase file size by a lot, which is why I think it should be a configurable option. When working with dev builds of a game, however, file size is not an issue, and its pretty easy to just run the maps through a minifier (this could even be done in CI) before releasing the game. Its far more trouble to have to run an external formatter every time I save a map.

**Comment by griest024** [comment]
@bjorn as for the usage of `pretty_generate` I just did this in `irb`.

```
$ irb
irb(main):001:0> require 'json'
=> true
irb(main):002:0> puts JSON.pretty_generate([0, 1, 2, 3, 4, 5])
[
  0,
  1,
  2,
  3,
  4,
  5
]
```

**Comment by griest024** [comment]
@termhn yes that's what I'm doing ATM, however it gets pretty annoying to have to run a formatter every time I save a file. I have to run it every time so that my saved file will diff correctly against the git version of the file.
